specs/agents/tracing-instrumentation-azure-functions.md

```diff
@@ -0,0 +1,103 @@
+Alternatively, agents could try to detect their hosting plan and not report cold starts for Premium plans at all.
+
+### Disabled Functionality
+
+Agents SHOULD detect whether they are running in an Azure Functions environment by testing
+for the existence of the `FUNCTIONS_WORKER_RUNTIME` environment variable.
+
+The following agent functionalities SHOULD to be turned off when tracing Azure Functions until decided otherwise:
+
+* **Metrics collection:** this includes all kind of metrics: system, process and breakdown metrics and is equivalent to
+setting `ELASTIC_APM_METRICS_INTERVAL = 0`
+* **Remote configuration:** equivalent to setting `ELASTIC_APM_CENTRAL_CONFIG = false`
+* **Cloud metadata discovery:**
+  * Equivalent to setting `ELASTIC_APM_CLOUD_PROVIDER = none`
+  * Instead an Azure Functions specific metadata detection MUST be performed (see see [Metadata](./metadata.md)).
+* **System metadata discovery:** in some agents, this may be a relatively heavy task. For example, the Java agent
+executes external commands in order to discover the hostname, which is not required for tracing Azure Functions. All other
+agents read and parse files to extract container and k8s metadata, which is not required as well.
+
+## Trigger-Specific Instrumentation
+
+Azure Functions can be triggered in [many different ways](https://learn.microsoft.com/en-us/azure/azure-functions/functions-triggers-bindings).
+A generic transaction for an Azure Functions invocation can be created independently of the actual trigger
+based on the [generic instrumentation](#generic-instrumentation) fields described above.
+
+However, depending on the trigger type, different information might be available that can be used
+to capture additional transaction data or that allows additional, valuable spans to be derived.
+
+### HTTP Trigger
+
+HTTP invocations typically provide *some* variants of an HTTP *request object* (e.g. [`HttpRequestData`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.identitymodel.protocols.httprequestdata))
+and an HTTP *response object* (e.g. [`HttpResponseData`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.azure.functions.worker.http.httpresponsedata)).
+
+Agents SHOULD use the information in the *request* and *response objects* to
+fill the HTTP context (`context.request` and `context.response`) fields in the same way it is done for HTTP transactions.
+
+In particular, agents MUST use HTTP headers to retrieve the `traceparent` and the `tracestate`
+and use those to start the transaction for the tracing the function execution.
+
+In addition the following fields should be set for HTTP trigger invocations:
+
+| Field | Value | Description | Source |
+| - | - | - | - |
+| `type` | `request`| Transaction type. Constant value for HTTP trigger invocations. | |
+| `name` | e.g. `GET /prod/proxy/{proxy+}` | Transaction name: HTTP method followed by a whitespace and the (resource) path. | *request object* |
```

Comment:
This example is from a Lambda triggered by API Gateway. If there is some real example that would be better to use.

The "req" object I see in a (locally run with `func`) JavaScript Azure Function with an HTTP trigger includes these fields. 

```
[2022-11-16T00:08:16.333Z]   req: {
[2022-11-16T00:08:16.334Z]     method: 'GET',
[2022-11-16T00:08:16.334Z]     url: 'http://localhost:7071/api/HttpExample?name=asdf',
[2022-11-16T00:08:16.334Z]     originalUrl: 'http://localhost:7071/api/HttpExample?name=asdf',
```

There is also some potential route pattern involved. E.g. https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-http-webhook-trigger?tabs=in-process%2Cfunctionsv2&pivots=programming-language-javascript#customize-the-http-endpoint shows an example HTTP binding configuration with URL parameters:

```json
{
    "bindings": [
    {
        "type": "httpTrigger",
        "name": "req",
        "direction": "in",
        "methods": [ "get" ],
        "route": "products/{category:alpha}/{id:int?}"
    },
```

In this case, I think we'd want the `transaction.name` to be "GET /.../projects/{category:alpha}/{id:int?}"  (where I'm not sure what, if anything, "/.../" should be.

Reply:
> This example is from a Lambda triggered by API Gateway. If there is some real example that would be better to use

Sorry, I copied this part over from the Lamba spec and obviously forgot to update the sample value here 🤦 .

> There is also some potential route pattern involved ...

That's valuable input. I'll add this to the spec.